main/worker.es6.js

```diff
@@ -6,133 +6,67 @@
-    pkg.ReleaseWebInterfaceJunction,
-    ctx);
-
-const daoCtx = ctx.__subContext__.createSubContext({
-  releaseDAO,
-  webInterfaceDAO,
-  releaseWebInterfaceJunctionDAO,
-});
-
-
-//
-// Setup services for page.
-//
-
-const apiMatrixController = pkg.ApiMatrixController.create(null, daoCtx);
-const apiConfluence = pkg.ApiConfluence.create({
-  apiVelocityDAO,
-  failureToShipDAO,
-  browserSpecificDAO,
-  aggressiveRemovalDAO,
-}, daoCtx);
-
-// Register API under known names.
-ctx.registry.register('apiMatrixController', null, foam.box.SkeletonBox.create({
-  data: apiMatrixController,
-}));
-ctx.registry.register('apiConfluence', null, foam.box.SkeletonBox.create({
-  data: apiConfluence,
-}));
-
-// Now that services are exported, begin accepting messages from page.
-onmessage = function(event) {
-  if (!event.data instanceof MessagePort) {
-    throw new Error('Unexpected control message', event.data);
+// Has initial "this is your name" message been received?
+let nameReceived = false;
+// Post-"this is your name" event backlog; worker context setup is async.
+let events = [];
+self.onmessage =  e => {
+  // If name already received, queue event in backlog.
+  if (nameReceived) {
+    events.push(e);
+    return;
   }
 
-  ctx.messagePortService.addPort(event.data);
-};
-
-//
-// Bind to services from page.
-//
+  nameReceived = true;
+
+  foam.assert(
+      e.data && e.data.name, 'Failed to recieve name from worker owner');
+  const name = e.data.name;
+
+  // Set up worker context.
+  //
+  // TODO(markdittmer): Code generation is really only necessary  for worker(s)
+  // that deal with generated.CompatData model. Should this be somehow
+  // conditional and/or driven by a signal from the worker's owner?
+  const compatClassURL = `${window.location.origin}/${pkg.DAOContainer.COMPAT_MODEL_FILE_NAME}`;
+  pkg.ClassGenerator.create({
+    classURL: compatClassURL,
+  }).generateClass().then(() => {
+    const ctx = foam.box.Context.create({
+      myname: name,
+      unsafe: false,
+      classWhitelist: require('../data/class_whitelist.json'),
+    }, pkg.DAOContainer.create());
+
+    // Replay or wait for message port connection from owner.
+    const portHandler = e => {
+      if (!e.data instanceof MessagePort) {
+        throw new Error('Unexpected control message', e.data);
+      }
 
-// Bind to page events service once service has established MessagePort
-// connection.
-const events = foam.box.PromisedBox.create({
-  delegate: new Promise(function(resolve, reject) {
-    ctx.messagePortService.connect.sub(function(_, __, namedBox) {
-      if (namedBox.name !== '/page') {
-        reject(new Error(`Unexpected MessagePort connection from:
-                              ${namedBox.name}`));
-        return;
+      // Sets self.onmessage.
+      ctx.messagePortService.addPort(e.data);
+    };
+    if (events.length > 0) {
+      portHandler(events[0]);
+      for (let i = 1; i < events.length; i++) {
```

Comment:
`for (let event of events)` will get you the same outcome without the closure problems of `for . in .`

Reply:
I used a counter because it starts at `1`, not `0`. I'll add a comment highlighting this.